libc/string_cpy.c

```diff
@@ -0,0 +1,752 @@
+	TEST_ASSERT_NOT_EMPTY(&ptrVal);
+	TEST_ASSERT_BYTES_EQUAL(ptr, ptrVal);
+	TEST_ASSERT_EQUAL_INT(ptr, ptrVal);
+
+	TEST_ASSERT_NOT_NULL(memcpy(&sizVal, &size, sizeof(size)));
+	TEST_ASSERT_NOT_EMPTY(&sizVal);
+	TEST_ASSERT_BYTES_EQUAL(size, sizVal);
+	TEST_ASSERT_EQUAL_DOUBLE(size, sizVal);
+
+	TEST_ASSERT_NOT_NULL(memcpy(&chrVal, &chr, sizeof(chr)));
+	TEST_ASSERT_NOT_EMPTY(&chrVal);
+	TEST_ASSERT_BYTES_EQUAL(chr, chrVal);
+	TEST_ASSERT_EQUAL_CHAR(chr, chrVal);
+}
+
+
+TEST(string_memcpy, copy_paste_special)
+{
+	char str[24] = { 0 };
+	char buff[PATH_MAX] = { 0 };
+	char strText[] = "Lorem ipsum \0 dolor";
+
+	/* Checking no sensitivity for NULL */
+	memset(str, 0, sizeof(str));
+	memcpy(str, strText, sizeof(strText));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(strText, str, 20);
+
+	/* Checking capability of handling big strings */
+	TEST_ASSERT_EQUAL_INT(1, big_string());
+	memcpy(buff, long_str, PATH_MAX);
+	TEST_ASSERT_EQUAL_STRING(buff, long_str);
+}
+
+
+/* 
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+ */
+
+
+TEST_SETUP(string_memccpy)
+{
+}
+
+
+TEST_TEAR_DOWN(string_memccpy)
+{
+}
+
+
+TEST(string_memccpy, copy_paste)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+
+#ifndef __phoenix__
+	int num = 12345678,
+		numVal = 0;
+	char bigBuff[PATH_MAX] = { 0 },
+		 str1[30] = { 0 };
+
+	char strPlaceholder[24] = MAECENAS_ID;
+
+	/* num in hex is equal BC614E */
+	memccpy(&numVal, &num, 0x61, sizeof(num));
+	TEST_ASSERT_NOT_EMPTY(&numVal);
+	TEST_ASSERT_EQUAL_HEX(0x614E, numVal);
+
+	memccpy(str1, LOREM_IPSUM, 111, sizeof(LOREM_IPSUM));
+	TEST_ASSERT_NOT_EMPTY(str1);
+	TEST_ASSERT_EQUAL_STRING("Lo", str1);
+
+	/* Combination of 2 strings after first sign "s" because of holding data outside range */
+	memccpy(strPlaceholder, str1, 'o', sizeof(strPlaceholder));
+	TEST_ASSERT_EQUAL_STRING("Loecenas id commodo", strPlaceholder);
+
+	memccpy(str1, CRAS_HENDERERIT, 'H', sizeof(str1));
+	TEST_ASSERT_EQUAL_STRING("Ho", str1);
+
+	/* Checking no sensitivity for NULL */
+	memccpy(bigBuff, &"This is new \0 string", 'g', PATH_MAX);
+	TEST_ASSERT_EQUAL_CHAR_ARRAY("This is new \0 string", bigBuff, 18);
+
+#endif
+}
+
+
+TEST(string_memccpy, copy_paste_special)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+
+#ifndef __phoenix__
+	char *buff[10], *bigBuff[PATH_MAX];
+	/* Checking correct pointer association */
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(buff, &"abc", 'h', sizeof("abc")));
+
+	TEST_ASSERT_EQUAL_PTR(&buff, memccpy(buff, &"abc", 'a', sizeof("abc")) - 1);
+
+	/* Ability to copy long string */
+	memccpy(bigBuff, &long_str, 'c', PATH_MAX);
+	TEST_ASSERT_EQUAL_STRING("aabc", bigBuff);
+
+	memccpy(bigBuff, &long_str, 'x', PATH_MAX);
+	TEST_ASSERT_EQUAL_STRING("x", strstr((const char *)bigBuff, "x"));
+#endif
+}
+
+
+/* 
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+ */
+
+
+TEST_SETUP(string_strncpy)
```

Comment:
Please apply the similar changes and split tests to more test groups in `strncpy` and `spncpy`. Please remember about the functionality you want to test in the written test group/test case. For example test group `string_strncpy, copy_paste` implies that copy paste utility will be tested, but in fact it's too general, we want to test for example for those functions above some basic stuff I mentioned above we would like to test:

Based on this requirement:
`(bytes that follow a NUL character are not copied)` - POSIX strncpy documentation

 - capability of copying the NUL terminating character: test case could be named `nul_term_copy`
 - not copying the character after NUL terminatin character: `nul_term_stop`

Based on this piece of doc:
`If the array pointed to by s2 is a string that is shorter than n bytes, NUL characters shall be appended to the copy in the array pointed to by s1, until n bytes in all are written`
 
we could test:

 - capability of appending NUL characters: test case named `append_nul`

and of course based on this line:
`The stpncpy() and strncpy() functions shall copy not more than n bytes ... from the array pointed to by s2 to the array pointed to by s1`

we could check whether:

- not more characters are copied, a test case called `no_more`
- different packages of data (various sizes, values) can be copied, so we could create cases like `basic` , `big`, `every_byte`, `one_byte`, `zero` and so on

It's only example, you don't have to follow all of those names and ideas, but think about it, what utility you want to test in this specific test case. Of course we sometimes do a few things in one test case, but then a name shall indicate it.